external-crates/move/crates/move-compiler/src/shared/ide.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{expansion::ast as E, parser::ast as P, shared::Name, naming::ast as N, typing::ast as T, ice, diagnostics::Diagnostics};
+
+use move_ir_types::location::Loc;
+
+use std::collections::BTreeMap;
+
+#[derive(Debug, Clone)]
+pub struct IDEInfo {
+    pub exp_info: BTreeMap<Loc, ExpEntry>
+}
+
+#[derive(Debug, Clone)]
+pub struct ExpEntry {
+    /// Location of the recorded info
+    pub loc: Loc,
+    /// Indicates this location was a macro call site
+    pub macro_call_info: Option<Box<MacroCallInfo>>,
+    /// Indicates this location is an expansion of a lambda
+    pub expanded_lambda: bool,
+}
+
+/// For use in the compiler, to ease recording IDE information.
+pub enum ExpInfo {
```

Comment:
It looks like `ExpEntry` and `ExpInfo` encode the same type of information. Can we just use one rather than using `ExpInfo` in the compiler and (at least seemingly) translating it almost immediately into `ExpEntry`?

Reply:
Unfortunately, this is likely the cleanest thing we can do. Since the actual information live on the `CompilationEnv` take a `Diagnostics` to report ICEs, the only alternative would be to re-expose the _entire_ interface for every information form on that environment, e.g., 

```
pub fun set_ide_exp_macro_call_info(&mut self, call_info: ide::MacroCallInfo) {
  ...
}

pub fun set_ide_exp_expanded_lambda(&mut self) {
  ...
}

pub fun set_ide_exp_autocomplete_dot_information(&mut self) {
  ...
}
```

The current approach avoids this by essentially baking this dispatch into the `ExpInfo`. I'm happy to change this if you want, but I think the overall usage and maintenance will become clumsier.